org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java

```diff
@@ -279,110 +274,130 @@ private static boolean capabilitiesComply(LanguageServerWrapper wrapper,
+		final var contentTypes = new ArrayDeque<IContentType>(LSPEclipseUtils.getDocumentContentTypes(document));
+		final var processedContentTypes = new HashSet<IContentType>();
+		final var path = new Path(uri.getPath());
+		final var file = LSPEclipseUtils.getFile(document);
 
-		synchronized (startedServers) {
-			// already started compatible servers that fit request
-			res.addAll(startedServers.stream()
-					.filter(wrapper -> {
-						try {
-							return wrapper.isConnectedTo(uri) || LanguageServersRegistry.getInstance().matches(document, wrapper.serverDefinition);
-						} catch (Exception e) {
-							LanguageServerPlugin.logError(e);
-							return false;
-						}
-					})
-					.filter(wrapper -> wrapper.canOperate(document))
-					.collect(Collectors.toList()));
+		while (!contentTypes.isEmpty()) {
+			final var contentType = contentTypes.poll();
+			if (contentType == null || processedContentTypes.contains(contentType)) {
+				continue;
+			}
 
-			while (!contentTypes.isEmpty()) {
-				IContentType contentType = contentTypes.poll();
-				if (contentType == null || processedContentTypes.contains(contentType)) {
+			for (ContentTypeToLanguageServerDefinition mapping : lsRegistry.findProviderFor(contentType)) {
+				if (mapping == null || !mapping.isEnabled()) {
 					continue;
 				}
-				for (ContentTypeToLanguageServerDefinition mapping : LanguageServersRegistry.getInstance()
-						.findProviderFor(contentType)) {
-					if (mapping == null || !mapping.isEnabled()) {
-						continue;
-					}
-					LanguageServerDefinition serverDefinition = mapping.getValue();
-					if (serverDefinition == null) {
-						continue;
-					}
-					if (startedServers.stream().anyMatch(wrapper -> wrapper.serverDefinition.equals(serverDefinition)
-							&& wrapper.canOperate(document))) {
-						// we already checked a compatible LS with this definition
-						continue;
-					}
-					final IProject fileProject = file != null ? file.getProject() : null;
-					LanguageServerWrapper wrapper = fileProject != null ? new LanguageServerWrapper(fileProject, serverDefinition) :
-						new LanguageServerWrapper(serverDefinition, path);
-					startedServers.add(wrapper);
-					res.add(wrapper);
+				final LanguageServerDefinition serverDefinition = mapping.getValue();
+				if (serverDefinition == null) {
+					continue;
+				}
+				final Predicate<LanguageServerWrapper> serverSelector2 = wrapper -> wrapper.serverDefinition
+						.equals(serverDefinition);
+				if (res.stream().anyMatch(serverSelector2)) {
+					// we already checked a compatible LS with this definition
+					continue;
 				}
-				if (contentType.getBaseType() != null) {
-					contentTypes.add(contentType.getBaseType());
+
+				synchronized (startedServers) {
+					// check again while holding the write lock
+					final var matchingServers = startedServers.stream() //
+							.filter(serverSelector) //
+							.filter(serverSelector2)
```

Comment:
can we name serverSelector and  serverSelector2 in a more meaningful way that is not 1 and 2?